cmd/aad-cli/cli/user.go

```diff
@@ -0,0 +1,220 @@
+	}
+
+	return users, cobra.ShellCompDirectiveNoFileComp
+}
+
+// getCache returns the cache, either from the options field if overridden or
+// a newly created one.
+func (a *App) getCache() (*cache.Cache, error) {
+	if a.options.cache != nil {
+		return a.options.cache, nil
+	}
+
+	return cache.New(a.ctx)
+}
+
+// getDefaultUser returns the current user name or a blank string if an error occurs.
+func getDefaultUser() string {
+	u, err := user.Current()
+	if err != nil {
+		return ""
+	}
+
+	return u.Username
+}
+
+// runUser executes a specific user action based on the arguments passed to the command.
+func runUser(ctx context.Context, args []string, c *cache.Cache, username string, allUsers bool, moveHome bool) error {
+	var err error
+	var key string
+	var value any
+
+	switch len(args) {
+	case 0:
+		// Return current user information or all user names if explicitly requested.
+		if allUsers {
+			var users []string
+			users, err = c.GetAllUserNames(ctx)
+			value = strings.Join(users, "\n")
+		} else {
+			var user cache.UserRecord
+			user, err = c.GetUserByName(ctx, username)
+			value, _ = user.IniString()
+		}
+	case 1:
+		// Return the value for the given key
+		key = args[0]
+		if key == "shadow_password" {
+			if !c.ShadowReadable() {
+				return fmt.Errorf("You do not have permission to read the shadow database")
+			}
+			var user cache.UserRecord
+			user, err = c.GetUserByName(ctx, username)
+			value = user.ShadowPasswd
+			break
+		}
+
+		value, err = c.QueryPasswdAttribute(ctx, username, key)
+	case 2:
+		// Set the value for the given key and exit
+		key, value = args[0], args[1]
+		if err := updateUserAttribute(ctx, c, username, key, value, moveHome); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	if err != nil {
+		return err
+	}
+
+	fmt.Println(strings.TrimSpace(fmt.Sprint(value)))
+	return nil
+}
+
+// updateUserAttribute updates the given attribute for an user to the specified value.
+// For some attributes such as home, additional actions are performed.
+func updateUserAttribute(ctx context.Context, c *cache.Cache, username, key string, value any, moveHome bool) error {
+	prevValue, err := c.QueryPasswdAttribute(ctx, username, key)
+	if err != nil {
+		return err
+	}
+
+	if prevValue == value {
+		logger.Debug(ctx, "No change to %q for %s", key, username)
+		return nil
+	}
+
+	if err := c.UpdateUserAttribute(ctx, username, key, value); err != nil {
+		return err
+	}
+
+	// Take additional actions based on the key that was updated
+	if key == "home" && moveHome {
+		// Update the home directory if it changed
+		if err := moveUserHome(ctx, username, fmt.Sprintf("%v", prevValue), fmt.Sprintf("%v", value)); err != nil {
+			return fmt.Errorf("Unable to move home directory for %s: %w", username, err)
+		}
+	}
+	return nil
+}
+
+// moveUserHome moves the home directory of an user from the previous location to the new one.
+func moveUserHome(ctx context.Context, username, prevValue, value string) error {
```

Comment:
I would add a defer func() {} here instead for `fmt.Errorf("Unable to move home directory for %s: %w", username, err) ` and just return the err directly in the function call.

This way, we are in line with the function itself decorating itself (if reused one day elsewhere)